Save Additional Metadata to SANS Reduced Output

cailafinn opened the pull request:
**Description of work**
Added back some metadata that had gone missing in the reduction output since Mantid 3.9.2.

**Purpose of work**
Retain metadata in the output needed for analysis down the road.

Requested by user @smk78 

**Summary of work**
- Adjusted SANSSave to make use of all of the saving parameters in SaveCanSAS1D
- Added those same properties to SaveNXCanSAS and use them in SANSSave
- Added loading for that metadata to LoadNXCanSAS

**To test:**

_Note: You may need to download a new version of the Training Course Data if you get errors about `Version 0` when loading the user file._

1. Open the SANS GUI
2. Load the user file and the batch file from the `loqdemo` from the Training Course Data
3. Save output to file (make sure your default mantid save directory is somewhere sensible)
5. Process All
6. Check your save directory and find the output file. Make sure the output is in a similar state to the before example in the issue. 

Fixes #28998 


---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

Commit:
Add thickness to CanSaS output

RE #28998

Commit:
Add geometry info to reduced file

RE #28998

Commit:
Include detectors in reduced output

RE #28998

Commit:
Add tests for get_detector_types helpers

RE #28998

Commit:
Add tests for getting detector names

RE #28998

Commit:
Remove accidentally added error update

RE #28998

Commit:
Add check for None geometry flag

RE #28998

Commit:
Add sample thickness to SaveNXCanSAS

RE #28998

Commit:
Add beam shape (geometry) to savenxcansas

RE #28998

Commit:
Add beam width & height to have nxcansas

RE #28998

Commit:
Output as doubles where appropriate

The data types for some of these make more sense to be doubles with
units. Switch them out and add some tests.

RE #28998

Commit:
Add a release note

RE #28998

Commit:
Save geometry flag as int in nexus output

RE #28998

Commit:
Fix SaveSANS converting to string on wrong algorithm

RE #28998

Commit:
Replace beam size with aperture gap sizes

On request from Steve at ISIS.

RE #28998

sf1919 commented:
@ConorMFinn  As there are what appears to be valid test failures we won't re-run the ones that are currently 'scheduled' as they will rerun when you push a new commit

Commit:
Store geometry in nx files as a string

These outputs are generally used for instrument debugging. So it makes
more sense for it to be a string rather than the index that we use to
keep track of it within the mantid sample objects. This requires some
manipulation of the values when loading, so new functions have been
added to allow for that.

There's also a fix here to make sure that we only try to open groups
that exist in the nexus file when loading, as this was causing crashes.

RE #28998

Commit:
Add test for saving SANS Sample to NeXus format

RE #28998

Commit:
Add comments and fix formatting

RE #28998

Commit:
Fix missed capital letter in geometry property

RE #28998

rbauststfc commented on Framework/DataHandling/src/SaveNXcanSAS.cpp:
```diff
@@ -722,26 +759,36 @@ void SaveNXcanSAS::init() {
   declareProperty("RadiationSource", "Spallation Neutron Source",
-                  std::make_shared<Kernel::StringListValidator>(radiation_source), "The type of radiation used.");
+                  std::make_shared<Kernel::StringListValidator>(radiationSourceOptions), "The type of radiation used.");
+  std::vector<std::string> geometryOptions{"Cylinder", "FlatPlate", "Disc"};
```
I'm still learning about C++ imports so this may not be a good suggestion, but now that we have constants for these geometry options defined in Sample.h would it be worth using them here?

rbauststfc commented on Framework/DataHandling/src/LoadNXcanSAS.cpp:
```diff
@@ -128,6 +129,62 @@ void loadLogs(H5::Group &entry, const Mantid::API::MatrixWorkspace_sptr &workspa
+// ----- SAMPLE
+
+namespace {
+bool hasAperture(H5::Group &entry) {
```
Would it be worth getting these methods that are checking if groups exist to return the group if it's found, so that we can avoid repeating the call to open the group again afterwards? I noticed the code further on uses `&&`, which I've read a bit about but am not very familiar with, so apologies if it should be obvious to me why this is more optimal!

Another C++ general info question, what were the reasons for defining these in a namespace rather than as private methods?

rbauststfc reviewed (commented):
I'm afraid this is just a partial review at this stage - I've done a quick manual test but need to do a little more with the extra data that we've been sent by the scientists. What I've checked so far is working well. I think there may be a slight issue with the geometry potentially always saving out as "Disc" (based on reading the code but not verified from manually testing) which I've flagged in one of my comments.

I also haven't managed to review a number of the test files yet, but I've finished looking at all of the functionality changes and thought it might be helpful to send you my review comments so far. Overall this is looking really good, and it's really great that you thought to update the loaders for NXcanSAS with the new information that's being saved out, good catch on that. I have just a few suggestions and C++ queries (mainly to help me learn a bit more myself, if that's OK!).

Commit:
Remove flag to shape conversion from save_workspace

RE #28998

ConorMFinn commented on scripts/SANS/sans/algorithm_detail/save_workspace.py:
```diff
@@ -52,14 +53,19 @@ def get_save_strategy(file_format_bundle, file_name, save_options, transmission_
         save_name = "SaveNexusProcessed"
     elif file_format is SaveType.CAN_SAS:
+        # Needs to be in string form for SaveCanSAS1D.
+        maybe_geometry = convert_to_shape(additional_properties["Geometry"])
```
Good catch! I think you're right about that hiding an issue as well. It's neater to have it in SANSSave anyway, I think

Commit:
Only set geometry if defined in sample.

Leave to the actual algorithms to decide the default.

RE #28998

Commit:
Replace detector names with an enum

RE #28998

Commit:
Simplify group opening for aperture and sample

RE #28998

rbauststfc commented on Framework/API/inc/MantidAPI/Sample.h:
```diff
@@ -37,6 +37,18 @@ class MANTID_API_DLL Sample {
+  /// Flag Value 1 = Cylinder
+  constexpr static char *CYLINDER{"Cylinder"};
+  /// Flag Value 2 = FlatPlate
+  constexpr static char *FLAT_PLATE{"Flat plate"};
```
I ran the following quick test to check saving out of the different geometry values using the data that Steve provided:

1) I ran the reduction.
2) Keeping the SANS dialog open (minimised), in the Python editor in Mantid I ran the following to change the geometry flag on the workspace to flat plate:

```
ws = mtd['104101_sans_nxs']
ws.sample().setGeometryFlag(2)
```

3) Then I went back to the SANS interface and ran the reduction again (clicked the Process button) and I got an error about `FlatPlate` not being an allowed Geometry property value.

I think the error happens because the enum returned by `sans.common.file_information.convert_to_shape` defines flat plate as `FlatPlate`, so I think we need to use the same here.

I also noticed that the `SampleShape` enum on the Python side doesn't define single crystal, so `convert_to_shape` will return None for a geometry flag of 4, so I think we need to extend the support to cover "SingleCrystal". If you repeat the above test using geometry flag 4 it currently silently saves the geometry as "Disc", which doesn't feel ideal so I wonder if it would be better to be letting the user know there's an issue with the geometry.

ConorMFinn replied:
So for some reason SANSBatchMode uses the `Flat plate` form rather than the `FlatPlate` form, so I wasn't certain of which one to use. 

You're definitely right about doing more to inform the user though. That should probably be a warning at least rather than letting it be silently changed.

rbauststfc replied:
For `FlatPlate` vs `Flat plate`, I think we need to be using the same value everywhere, as if I'm following correctly than the code in `SANSBatchMode.py` is trying to pass geometry properties to save algorithms as well so I would expect this to produce an error too. If we use the enum everywhere then that might help maintain some consistency.

rbauststfc replied:
Having said that, looking at the original validator values for the `SaveCanSAS1D` geometry property, I think we want to stick with what was originally in use (and I assume Steve hasn't asked us to start supporting Single Crystal). So on line 168 of `SANSSave.py` I think we need to be assigning the value "Flat plate" rather than the value we get from the enum. Given how confusing this is at the moment, it might be good if we could define the geometry values in a re-usable way on the Python side rather than have string literals in multiple places.

ConorMFinn replied:
I agree. I wonder if we can export the values from Sample.h to python and use them from there. That might be the best way to maintain consistency

rbauststfc replied:
That would be ideal if it was possible - definitely worth a look, thanks.

ConorMFinn replied:
I had a bit more of a look around with this and it seems that the `FlatPlate` form is more commonly used in the project and `Flat plate` is generally only used by the SANS interface. 

From what Steve has told me, the sample information is intended just to be read rather than processed by any kind of data analysis process. So, as long as it's in a readable format and is consistent I think we can go with the one more commonly used and solidify the format now. 

My ideal would be to have everything from here on be saved as `FlatPlate` but allow our loading algorithms to convert `Flat plate` to the right geometry flag at load time for the sake of backwards compatibility.

rbauststfc commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSSave.py:
```diff
@@ -134,6 +137,16 @@ def PyExec(self):
         file_name = self.getProperty("Filename").value
         workspace = self.getProperty("InputWorkspace").value
 
+        sample = workspace.sample()
```
While I was testing I realised that you can change the thickness, shape, height and width values in the batch table of the SANS interface. I'm afraid I'm not familiar enough with the interface or code to know how this is intended to be used, but I noticed that because we're taking these values from the workspace when we're saving it means that any changes to the batch table are not saved out. Is this right, or should we be saving whatever is in the batch table?

rbauststfc reviewed (commented):
Thanks very much for the changes made so far. There are just a few more bits I noticed during testing and re-reviewing that would be worth a look, sorry.

ConorMFinn commented:
There has also been a note from Steve. Mainly that the sample information (such has thickness and beam shape) should be taken from the interface rather than the loaded nexus file. Doing some digging made me realise that I had incorrectly assumed that the sample object in the output workspace would be automatically updated by the reduction. Which it isn't, so changing that to take the information from the gui state and apply it to the output workspace before starting saving needs to be done before this is merged.

rbauststfc commented:
Good to have confirmation from Steve on where the values should be coming from. On taking the information from the GUI state and applying it to the output workspace, I can see the logic in having the output workspace match the values used in the reduction, but just to double check has Steve mentioned whether it would be OK/accurate for us to update these values in the output workspace?

ConorMFinn commented on Framework/DataHandling/inc/MantidDataHandling/NXcanSASDefinitions.h:
```diff
@@ -79,6 +79,17 @@ const std::string nxInstrumentSourceClassAttr = "NXsource";
 
 const std::string sasInstrumentSourceGroupName = "sassource";
 const std::string sasInstrumentSourceRadiation = "type";
+const std::string sasInstrumentSourceBeamShape = "beam_shape";
```
Ah, I think they're a hangover from before I moved the data over to being stored in `Aperture`, after getting some feedback from Steve

ConorMFinn commented:
> has Steve mentioned whether it would be OK/accurate for us to update these values in the output workspace?

I just checked to make sure, and yes. The values put into the table should override the values brought in from the nexus file and so should be present in the output workspaces and files.

rbauststfc commented on Framework/DataHandling/src/SaveNXcanSAS.cpp:
```diff
@@ -722,26 +760,36 @@ void SaveNXcanSAS::init() {
-                  std::make_shared<Kernel::StringListValidator>(radiation_source), "The type of radiation used.");
+                  std::make_shared<Kernel::StringListValidator>(radiationSourceOptions), "The type of radiation used.");
+  std::vector<std::string> const geometryOptions{Sample::CYLINDER, Sample::FLAT_PLATE, Sample::DISC};
+  declareProperty("Geometry", std::string(Sample::DISC), std::make_shared<Kernel::StringListValidator>(geometryOptions),
```
I've been thinking about this new `Geometry` property and backwards compatibility. If someone had used `SaveNXcanSAS` in a script that they had written using a previous version of Mantid and then ran that script in a version of Mantid that contains these changes, their saved file will contain a geometry entry set to the default of `Disc` where previously it wouldn't have contained any information. Similarly I guess it might also affect re-running history depending on what someone has done with a workspace. I'm wondering if we should consider making this parameter optional/setting the default to an empty string and only saving out geometry information if it's been passed in?

ConorMFinn replied:
I've asked Steve about this. He's suggested that we have some other value such as `Unknown` as the value should be set on the file by IBEX when making the run. I'm not sure how this fits in with our sample objects.

rbauststfc commented on Framework/DataHandling/src/SaveNXcanSAS.cpp:
```diff
@@ -722,26 +760,36 @@ void SaveNXcanSAS::init() {
+  std::vector<std::string> const geometryOptions{Sample::CYLINDER, Sample::FLAT_PLATE, Sample::DISC};
+  declareProperty("Geometry", std::string(Sample::DISC), std::make_shared<Kernel::StringListValidator>(geometryOptions),
+                  "The geometry type of the collimation.");
+  declareProperty("SampleHeight", 0.0,
```
I recall adding new parameters to an algorithm in the middle of the parameter list once and it was pointed out to me that this will break any scripts that are using positional arguments. Would it be better to add these new parameters to the end of the list? If moving them is an issue then I think we could also make sure that the parameter change is really clearly highlighted in the release notes.

rbauststfc reviewed (commented):
Sorry, I know this is a work in progress but there are a couple of things I keep meaning to mention about the new `SaveNXcanSAS` parameters, in case it helps.

Commit:
Avoid positional argrument changes

Move the new properties to the end, this will stop them from messing up
the positional arguments when this algorithm is called in a script.

RE #28998

Commit:
Fix output workspaces not having correct sample info

RE #28998

rbauststfc commented:
Just adding a comment to confirm that this is still a work in progress.

Commit:
Handle multiple geometry string formats

RE #28998

Commit:
Remove unused NX definitons

RE #28998

Commit:
Add option for unknown geometry flag

Given this can be any kind of weird string and we have to do awkward
conversions, make sure that we don't automatically convert to disc or
just be blank and make it look like the metadata was just discarded
rather than explicitly unknown.

RE #28998

rbauststfc reviewed (commented):
The code changes are looking good and values entered into the GUI are now correctly being picked up in the workspace and saved out.

I'm still getting an issue with the `FlatPlate` geometry for the `SaveCanSAS1D` algorithm, though. If you want to change the sample shape in the GUI batch table to flat plate then you have to enter `FlatPlate`. It's this syntax for the geometry flag that we use in `convert_to_shape` when we're getting the geometry name from the integer value in the workspace for saving, however `SaveCanSAS1D` only allows `Flat Plate` and so you get a save error. Everything works OK for `SaveNXcanSAS`, so perhaps we just need to add `FlatPlate` as one of the supported geometry property values for `SaveCanSAS1D`?

ConorMFinn commented on Framework/DataHandling/src/LoadNXcanSAS.cpp:
```diff
@@ -128,6 +130,68 @@ void loadLogs(H5::Group &entry, const Mantid::API::MatrixWorkspace_sptr &workspa
+      sample.setWidth(width.front());
+    }
+    auto &&geometry = Mantid::DataHandling::H5Util::readString(apertureGroup, sasInstrumentApertureShape);
+    boost::to_lower(geometry);
```
`switch`es only work for `int`s in C++, so we'd need an enum of the various options, which is something that I'd like to iron out in a later issue where we make geometry naming consistent across the codebase. 

For now, `if`/`else` will work well though, nice catch.

Commit:
Make NX and CanSAS1D saving and loading consistent

We should always be saving the geometry as the same strings in both file
types, but we need to be able to load all reasonable permutations of the
string for legacy reasons. In SANS, generally FlatPlate should be used
where posssible.

RE #28998

ConorMFinn commented on Framework/DataHandling/src/LoadCanSAS1D.cpp:
```diff
@@ -34,16 +35,15 @@ using namespace Mantid::API;
 
 namespace {
-int getGeometryID(const std::string &selection) {
+int getGeometryID(std::string selection) {
```
We can't guarantee `constness` in this method anymore as `tolower` modifies the string in place. As a result, passing the value by reference would introduce hidden behaviour as it would get you the geometry ID, but it would also modify the string you passed in to make it lowercase unless we made a copy inside anyway. 

Neither of these are ideal, so setting the `selection` string to be passed by value means that you are forced to use `std::move` at the call point. You therefore give up your right to use it in the caller, and we can modify it in `getGeometryID()` without worrying about it causing issues elsewhere.

rbauststfc replied:
Ah, thanks, that makes sense - I hadn't thought properly about the effect of the conversion to lowercase.

ConorMFinn commented on Framework/DataHandling/src/SaveCanSAS1D.cpp:
```diff
@@ -101,7 +101,7 @@ void SaveCanSAS1D::init() {
   std::vector<std::string> collimationGeometry{
       "Cylinder",
-      "Flat plate",
+      "FlatPlate",
```
I think that having the default as `Disc` is more prone to significant errors than having the default as `Unknown`. I believe that Steve suggested we have something like that instead anyway. I think as long as we make it clear enough in the release notes, this is still a change worth making.

rbauststfc replied:
OK, if we add something to the release notes to make sure that change is clear (and we can also flag it up to the SANS group specifically) then that sounds good.

Commit:
Add an extra release note for clarity.

RE #28998

rbauststfc reviewed (approved):
Thanks very much for sticking with it to get this one all sorted Conor. The code is all looking good and I've given things a thorough test. The saving is working well, the missing information has now been restored, values are picked up from the GUI as expected and the saved file can be successfully loaded back into Mantid.

Gatekeepers, I'm happy that all previous review comments have been addressed and this is ready to go in.
